hydra/hydra.go

```diff
@@ -192,7 +205,73 @@ func NewHydra(ctx context.Context, options Options) (*Hydra, error) {
 		go handleBootstrapStatus(hdCtx, bsCh)
 
 		hds = append(hds, hd)
+
+		// dial back hooks
+		evts := []interface{}{
+			new(event.EvtPeerIdentificationCompleted),
+		}
+
+		subs, err := hd.Host.EventBus().Subscribe(evts, eventbus.BufSize(256))
+		if err != nil {
+			return nil, fmt.Errorf("head could not subscribe to eventbus events; err: %w", err)
+		}
+		go func(hd *head.Head, subs event.Subscription) {
+			defer subs.Close()
+			for {
+				select {
+				case evt := <-subs.Out():
+					ev := evt.(event.EvtPeerIdentificationCompleted)
+					p := ev.Peer
+					addrs := hd.Host.Peerstore().Addrs(p)
+
+					// dial back on quic if peer advertises a quic address
+					for _, a := range addrs {
+						if mafmt.QUIC.Matches(a) {
+							nQUICConns.Add(1)
+
+							if err := hd.QuicDialBackHost.Connect(hdCtx, peer.AddrInfo{ID: p}); err == nil {
```

Comment:
We're keeping track of unique peer IDs already, could we somehow use that info to not attempt to dial back to a peer we've already dialed back to?

Reply:
@alanshaw I thought about it and it can really tricky. More importantly, what if a peers NAT/router etc has changed and a dial that was failing earlier would work if we re-tried ? Hence, I'd like to keep this simple without making any assumptions about the connectivity status of the peers.

Reply:
Gotcha, ok.